### PR TITLE
Base auto-capture logic on yield calls in template source

### PR DESCRIPTION
**Pull request by kaspth**
Extracted from #44

Currently, we're basing our auto-capture logic on whether there's a
single argument passed to render, e.g.:

```ruby
render "card" do |partial|
  partial.yield :title, t(".title")
end
```

But if users want to pass extra content via an explicit `<%= yield something %>`,
we don't support that.

This changes our logic to look for `<%= yield something %>` calls in the
template source and not auto-capture in that case. We will pass the partial
on as an extra argument in case callers want to expose that as a named block
argument.

We'll still auto-capture in case a template has non-capturing yields like
`<%= yield :title %>`.

### CODE REVIEW

**Comment by kaspth** [comment]
Note: this also relies on some code structure in `ActionView::PartialRenderer` that was added in 6.1. I don't know how tied we are to backwards compatibility, but with the `_layout_for` override users would just need to insert a plain `<%= yield %>` into their partials, or we could look at explicit versions support for Rails 5.2 and possibly further down. What do you say @andrewculver?

**Comment by andrewculver** [comment]
@kaspth No need to provide backward compatibility. Happy to support 6.1 and up.

File: lib/nice_partials/monkey_patch.rb [hunk 1]
```diff
@@ -40,13 +46,44 @@ def p(*args)
...
+  #   <%= yield something_else %>
+  #
+  # Doesn't match obfuscated `content_for` invocations, nor custom yields:
+  #   <%= yield :message %>
+  #   <%= something.yield %>
+  #
+  # Note: `<%= yield %>` becomes `yield :layout` with no `render` `block`, though this method assumes a block is passed.
+  def has_capturing_yield?
+    defined?(@has_capturing_yield) ? @has_capturing_yield :
+      @has_capturing_yield = source.match?(/\byield[\(? ]+(%>|[^:])/)
```
Comment (domchristie) [inline lib/nice_partials/monkey_patch.rb:87]: Just noticing the `%>`: does this work with HAML?

Comment (kaspth) [reply lib/nice_partials/monkey_patch.rb:87]: Originally I wanted to define this after template compilation, so we'd be Template Handler independent, but I just realized we'd need to trigger compilation for that to work (and that's why it didn't work).

I've never used anything besides ERB, how does HAML syntax go by the way? Like this?

```haml
%= yield
%= yield :title
```

Do you know what Slim looks like?

Comment (domchristie) [reply lib/nice_partials/monkey_patch.rb:87]: I'm not a HAML/Slim user, but the `%>` got me thinking.

> I've never used anything besides ERB, how does HAML syntax go by the way? Like this?

Yep I think so.

Slim looks simlar http://slim-lang.com/

**Pull request merged**